control-plane/p4RuntimeSerializer.cpp

```diff
@@ -639,6 +640,55 @@ getMatchType(cstring matchTypeName) {
     }
 }
 
+/*
+ * This function supports a deeply nested mix of Type_Newtype and typedef.
+ * The function also supported Type_Bits.
+ * A non-zero width, or zero width, or -1 is returned by midWidthbits() used
+ * by this function. For -1, a failure code, the BUG_CHECK is invoked.
+ * Therefore the uint32_t returned by the function is fine.
+*/
+static int
+getTypeWidth(const IR::Type* type, TypeMap* typeMap) {
+    return typeMap->minWidthBits(type, nullptr);
+}
+
+/*
+ * The function returns a cstring for use as type_name for any nested type.
+*/
+static cstring
+getTypeName(const IR::Type* type, std::vector<const IR::Type_Newtype*>& list,
```

Comment:
Any MatchField entry can be a deeply-nested `type`.  This is the reason the function recurses.  The list is needed because once the underlying type is reached, the front of the list has the type that is used for type_name.  For each MatchField, the list is preserved during recursion to yield the type_name needed.  Then, for the next MarchField entry, of course, the vector is cleared so that the next MatchField nested type can be processed.

Reply:
We don't care about all the types in the chain, we only care about the *first* type for the name, which is what `getTypeType` will give you. Explain to me where you are using the deeply-nested types, because all I'm seeing is `list.front()` (the first type) being returned and everything else being discarded.

Reply:
I agree the first block of code related to Type_Bits can be removed.  I had this code when I had to match annotation and only if annotated, I would generate a type_name.  However, Andy, told me, no annotation check is required for type_name generation.

Regarding rest of the code, I think, once we get agreement on what is generated, it would be easy to finalize the code. 

The following nested types exist in the test P4 program and are used for MatchField. 
One rule being followed is, if any MatchField contains a "T", then, a type_name is generated.  

The test program is: `testdata/p4_16_samples/p416-type-use3.p4`.  Just search for "key" in this program and you will find the table key being used for MatchFields.

Please take a look at the generated p4info and let me know if the type_names are correct and which ones are not.  thanks.

`testdata/p4_16_samples/p416-type-use3.p4.p4info.txt`

```
// Since types and typedefs can be defined in terms of each other, the
// names I use here contain sequences of Ds and Ts to indicate the
// order in which they have been "stacked", e.g. EthDT_t is a `type`
// (the T is last) defined on type of a `typedef` (the D just before
// the T).

typedef bit<48> EthD_t;
@p4runtime_translation("p4.org/psa/v1/EthT_t", 48)
type    bit<48> EthT_t;

typedef bit<8>     CustomD_t;
type    bit<8>     CustomT_t;
typedef CustomD_t  CustomDD_t;
type    CustomD_t  CustomDT_t;
typedef CustomT_t  CustomTD_t;
type    CustomT_t  CustomTT_t;
typedef CustomDD_t CustomDDD_t;
type    CustomDD_t CustomDDT_t;
typedef CustomDT_t CustomDTD_t;
type    CustomDT_t CustomDTT_t;
typedef CustomTD_t CustomTDD_t;
type    CustomTD_t CustomTDT_t;
typedef CustomTT_t CustomTTD_t;
type    CustomTT_t CustomTTT_t;
```

Reply:
I agree that no check for a `p4runtime_translation` annotation should be needed to determine the value of a `type_name` in the P4Info file for table search key fields, action parameters, value_set match fields, and controller_header fields.

It _will_ be needed at some point in order to determine the value of the `bitwidth` field in the P4Info file for those things.